torch_xla/distributed/fsdp/xla_fully_sharded_data_parallel.py

```diff
@@ -1471,48 +1478,58 @@ def get_original_names_and_sharded_parameters(self):
 
     return orig_named_parameters
 
-  def get_shard_metadata(self):
+  def load_state_dict(self, state_dict, strict=True):
+    if "shard_metadata" in state_dict:
+      state_dict.pop("shard_metadata")
+    super().load_state_dict(state_dict, strict=strict)
+
+  @staticmethod
+  def get_shard_metadata(m, destination, prefix, local_metadata):
     """
     Get the shard metadata to consolidate the sharded model checkpoints.
-    The output from this method should be saved in a checkpoint file and
-    used in ``consolidate_sharded_model_checkpoints``.
+    The output from this method will be used in ``consolidate_sharded_model_checkpoints``.
     """
     shard_info = {}
     flatten_info = {}
     buffer_info = {}
-    for module_name, m in self.named_modules():  # includes self
-      # remove "_fpw_module." from module names since it is also removed in
-      # XlaFullyShardedDataParallel's state_dict()
-      module_name = module_name.replace("_fpw_module.", "")
-
-      if isinstance(m, XlaFullyShardedDataParallel):
-        sharded_param_info = {}
-        for p_shard in m.sharded_params:
-          sharded_param_info[p_shard._name] = {
-              "_orig_size": p_shard._orig_size,
-              "_orig_name": p_shard._orig_name,
-          }
-        shard_info[module_name] = sharded_param_info
-
-      if isinstance(m, XlaFlattenParamsWrapper):
-        for i in range(len(m.flat_params)):
-          param_name = f"flat_param_{i}"
-          if module_name != "":
-            param_name = module_name + "." + param_name
-          flatten_info[param_name] = m.metadata(i)
-
-    for name, buf in self.named_buffers():
+    # remove "_fpw_module." from module names since it is also removed in
+    # XlaFullyShardedDataParallel's state_dict()
+    module_name = prefix.replace("_fpw_module.", "")
+    # remove the trailing "." since state_dict will add it
+    module_name = module_name.rstrip(".")
+
+    if isinstance(m, XlaFullyShardedDataParallel):
+      sharded_param_info = {}
+      for p_shard in m.sharded_params:
+        sharded_param_info[p_shard._name] = {
+            "_orig_size": p_shard._orig_size,
+            "_orig_name": p_shard._orig_name,
+        }
+      shard_info[module_name] = sharded_param_info
+
+    if isinstance(m, XlaFlattenParamsWrapper):
+      for i in range(len(m.flat_params)):
+        param_name = f"flat_param_{i}"
+        if module_name != "":
+          param_name = module_name + "." + param_name
+        flatten_info[param_name] = m.metadata(i)
+
+    for name, buf in m.named_buffers(recurse=False):
```

Comment:
If we have `recurse=False`, would we miss buffers such as `model.module1.module2.buffer` when we only wrap the base `model` with the FSDP class?